compiler/rustc_infer/src/infer/outlives/verify.rs

```diff
@@ -82,27 +83,39 @@ impl<'cx, 'tcx> VerifyBoundCx<'cx, 'tcx> {
         debug!("param_bound(param_ty={:?})", param_ty);
 
         // Start with anything like `T: 'a` we can scrape from the
-        // environment
-        let param_bounds = self
-            .declared_generic_bounds_from_env(GenericKind::Param(param_ty))
-            .into_iter()
-            .map(|outlives| outlives.1);
+        // environment. If the environment contains something like
+        // `for<'a> T: 'a`, then we know that `T` outlives everything.
```

Comment:
So this is the type of thing that worries me for forward-compatibility. If we eventually mean that `'a` to only mean lifetimes where T is WF, then `'a` doesn't outlive *everything*, necessarily.

(I'm just reading comments for now, not code)

Reply:
This code applies to `ParamTy`s only and they're assumed to be WF unconditionally, so i see no problem here.

Reply:
I don't think that's a problem. This only occurs when there is a where-clause in the environment that says `for<'a> T: 'a`. That will always mean *everything*. The situation you're talking about is different -- e.g., `for<'a> <T as Foo<'a>>::Item: 'b`.